mm2src/mm2_main/src/lp_swap/swap_watcher.rs

```diff
@@ -57,204 +54,235 @@ pub struct TakerSwapWatcherData {
-                secret: H256Json::default(),
-            }),
-            data,
+#[async_trait]
+impl State for ValidateTakerFee {
+    type Ctx = WatcherContext;
+    type Result = ();
+
+    async fn on_changed(self: Box<Self>, watcher_ctx: &mut WatcherContext) -> StateResult<Self::Ctx, Self::Result> {
+        let validated_f = watcher_ctx
+            .taker_coin
+            .watcher_validate_taker_fee(
+                watcher_ctx.data.taker_fee_hash.clone(),
+                watcher_ctx.verified_pub.clone(),
+            )
+            .compat();
+        if let Err(e) = validated_f.await {
+            Self::change_state(Stopped::from_reason(StopReason::ValidateTakerFeeFailed(
+                ERRL!("!watcher.watcher_validate_taker_fee: {}", e).into(),
+            )));
         }
+        Self::change_state(ValidateTakerPayment {})
     }
+}
 
-    async fn start(&self) -> Result<(Option<WatcherCommand>, Vec<WatcherEvent>), String> {
-        Ok((Some(WatcherCommand::ValidateTakerPayment), vec![WatcherEvent::Started]))
-    }
+#[async_trait]
+impl State for ValidateTakerPayment {
+    type Ctx = WatcherContext;
+    type Result = ();
+
+    async fn on_changed(self: Box<Self>, watcher_ctx: &mut WatcherContext) -> StateResult<Self::Ctx, Self::Result> {
+        let search_input = WatcherSearchForSwapTxSpendInput {
+            time_lock: watcher_ctx.data.taker_payment_lock as u32,
+            taker_pub: &watcher_ctx.data.taker_pub,
+            maker_pub: &watcher_ctx.data.maker_pub,
+            secret_hash: &watcher_ctx.data.secret_hash,
+            tx: &watcher_ctx.data.taker_payment_hex,
+            search_from_block: watcher_ctx.data.taker_coin_start_block,
+            swap_contract_address: &None,
+            swap_unique_data: &[],
+        };
 
-    // Do we need the exact same validation as the maker, or should we use a simpler validation process?
-    async fn validate_taker_payment(&self) -> Result<(Option<WatcherCommand>, Vec<WatcherEvent>), String> {
-        let wait_duration = (self.data.lock_duration * 4) / 5;
-        let wait_taker_payment = self.data.swap_started_at + wait_duration;
-        let confirmations = min(self.data.taker_payment_confirmations, TAKER_SWAP_CONFIRMATIONS);
+        match watcher_ctx
+            .taker_coin
+            .watcher_search_for_swap_tx_spend(search_input)
+            .await
+        {
+            Ok(Some(FoundSwapTxSpend::Spent(_))) => {
+                return Self::change_state(Stopped::from_reason(StopReason::TakerPaymentAlreadySpent))
+            },
+            Ok(Some(FoundSwapTxSpend::Refunded(_))) => {
+                return Self::change_state(Stopped::from_reason(StopReason::TakerPaymentAlreadyRefunded))
+            },
+            Err(err) => {
+                return Self::change_state(Stopped::from_reason(StopReason::TakerPaymentSearchForSwapFailed(
+                    ERRL!("!watcher.wait_for_confirmations: {}", err).into(),
```

Comment:
It seems error message should be `!watcher_search_for_swap_tx_spend.wait_for_confirmations:`. Also, it's preferred to avoid using `ERRL!` and other macros as we are trying to replace them with `MmError` everywhere: https://github.com/KomodoPlatform/atomicDEX-API/issues/1250.